mesonbuild/scripts/dist.py

```diff
@@ -0,0 +1,123 @@
+        for d in dirs:
+            dname = os.path.join(root, d)
+            zf.write(dname, dname[removelen:])
+        for f in files:
+            fname = os.path.join(root, f)
+            zf.write(fname, fname[removelen:])
+    zf.close()
+
+def create_dist(dist_name, src_root, bld_root, dist_sub):
+    distdir = os.path.join(dist_sub, dist_name)
+    gitdir = os.path.join(src_root, '.git')
+    if os.path.exists(distdir):
+        shutil.rmtree(distdir)
+    os.makedirs(distdir)
+    dest_gitdir = os.path.join(distdir, '.git')
+    shutil.copytree(gitdir, dest_gitdir)
+    subprocess.check_call(['git', 'reset', '--hard'], cwd=distdir)
+    shutil.rmtree(dest_gitdir)
+    for f in glob(os.path.join(distdir, '.git*')):
+        os.unlink(f)
+    xzname = distdir + '.tar.xz'
+    zipname = distdir + '.zip'
+    # Should use shutil but it got xz support only in 3.5.
+    tf = tarfile.open(xzname, 'w:xz')
+    tf.add(distdir, os.path.split(distdir)[1])
+    tf.close()
+    create_hash(xzname)
+    create_zip(zipname, distdir)
+    create_hash(zipname)
+    return (xzname, zipname)
+
+def check_dist(packagename):
+    print('Testing distribution package %s.' % packagename)
+    unpackdir = tempfile.mkdtemp()
+    builddir = tempfile.mkdtemp()
+    installdir = tempfile.mkdtemp()
+    ninja_bin = detect_ninja()
+    try:
+        tf = tarfile.open(packagename)
+        tf.extractall(unpackdir)
+        srcdir = glob(os.path.join(unpackdir, '*'))[0]
+        mesonbin = os.path.join(os.path.dirname(os.path.abspath(__file__)), '../..', 'meson.py')
+        if subprocess.call([sys.executable, mesonbin, srcdir, builddir]) != 0:
+            print('Running Meson on distribution package failed')
+            return 1
+        if subprocess.call([ninja_bin], cwd=builddir) != 0:
+            print('Compiling the distribution package failed.')
+            return 1
+        if subprocess.call([ninja_bin, 'test'], cwd=builddir) != 0:
+            print('Running unit tests on the distribution package failed.')
+            return 1
+        myenv = os.environ.copy()
+        myenv['DESTDIR'] = installdir
+        if subprocess.call([ninja_bin, 'install'], cwd=builddir, env=myenv) != 0:
+            print('Installing the distribution package failed.')
+            return 1
+    finally:
+        shutil.rmtree(srcdir)
+        shutil.rmtree(builddir)
+        shutil.rmtree(installdir)
+    print('Distribution package %s tested.' % packagename)
+    return 0
+
+def run(args):
+    src_root = args[0]
+    bld_root = args[1]
+    priv_dir = os.path.join(bld_root, 'meson-private')
+    dist_sub = os.path.join(bld_root, 'meson-dist')
+
+    buildfile = os.path.join(priv_dir, 'build.dat')
+
+    build = pickle.load(open(buildfile, 'rb'))
+    
+    dist_name = build.project_name + '-' + build.project_version
+
+    if not os.path.isdir(os.path.join(src_root, '.git')):
+        print('Dist currently only works with Git repos.')
```

Comment:
Git is the 99% solution. We did the same with wraps. Originally it only supported Git but people using Mercurial sent patches to add support.